File: php/union_find/union_find_sz.php
<?php
/**
 * 并查集sz优化版
 * 添加一个sz字典来统计此节点下级有多少个节点，在并操作中，节点少的顶级节点指向节点多的顶级节点
 */
class UnionFindSz{
    public $root = array();
    public $sz = array();
    public $count = 0;

    public function __construct($init_list = array()){
        $this->count = count($init_list);
        foreach($init_list as $v){
            $this->root[$v] = $v;
            $this->sz[$v] = 1;
        }
    }

    /**
     * 查操作
     *
     * @param int $key 键值
     * @return void
     */
    public function find($key){
        if($key < 0 || $key >= $this->count){
            return null;
        }
        while($this->root[$key] != $key){
            $key = $this->root[$key];
        }
        return $key;
    }

    /**
     * 判断两键值是否连接
     *
     * @param int $p 键值
     * @param int $q 键值
     * @return void
     */
    public function isConnected($p, $q){
        return $this->find($q) == $this->find($p);
    }

    /**
     * 并操作
     * 找出两键值的顶级节点；
     * 判断是否相同，如果相同则不操作；
     * 判断两顶级节点的子节点数量，数量少的顶级节点指向数量多的顶级节点；
     * 更新顶级节点的子节点数；
     *
     * @param int $p 键值
     * @param int $q 键值
     * @return void
     */
    public function unionElements($p, $q){
        $p_root = $this->find($p);
        $q_root = $this->find($q);
        if($q_root == $p_root){
            return;
        }
        if($this->sz[$p_root] > $this->sz[$q_root]){
            $this->root[$q_root] = $p_root;
            $this->sz[$p_root] += $this->sz[$q_root];
        }else{
            $this->root[$p_root] = $q_root;
            $this->sz[$q_root] += $this->sz[$p_root];
        }
    }
}

File: python/union_find/union_find_sz.py
class UnionFindSz:
    '''并查集优化方案一
    创建sz字典来记录根节点的子节点总数
    '''
    root = dict()
    sz = dict()
    count = 0

    def __init__(self, init_list=[]):
        '''初始化
        字典的键为元素值，字典的值为此元素值的父节点；
        所有节点都是根节点，并且无子节点，子节点数统计为1(或0)
        '''
        self.count = len(init_list)
        for i in init_list:
            self.root[i] = i
            self.sz[i] = 1

    def find(self, key):
        '''查操作'''
        if key < 0 and key >= self.count:
            return None
        while self.root[key] != key:
            key = self.root[key]
        return key

    def isConnected(self, p, q):
        '''判断两键值是否连接'''
        return self.find(p) == self.find(q)

    def unionElements(self, p, q):
        '''并操作
        查询两键值的顶级父节点，如果相同则不操作；
        判断两键值的顶级父节点的子节点数，子节点数少的指向子节点数多的(子节点数多的做顶级节点)，如果子节点数相同，则任意指向；
        注意，并操作后，需要将新加入的子节点数量添加到当前顶级节点上；
        '''
        p_root = self.find(p)
        q_root = self.find(q)
        if q_root == p_root:
            return
        if self.sz[p_root] > self.sz[q_root]:
            self.root[q_root] = self.root[p_root]
            self.sz[p_root] += self.sz[q_root]
        else:
            self.root[p_root] = self.root[q_root]
            self.sz[q_root] += self.sz[p_root]


if __name__ == "__main__":
    init_list = []
    list_len = 5
    for i in range(0, list_len):
        init_list.append(i)

    print(init_list)
    fund = UnionFindSz(init_list)

    print(fund.root)
    fund.unionElements(2, 3)
    fund.unionElements(3, 4)
    fund.unionElements(4, 0)
    print(fund.root)
    print(fund.sz)
